### PR TITLE
Dispatch exactly same text string for all slices.

### PR DESCRIPTION
**Pull request by eurekaka**
Include a map from sliceIndex to gang_id in the dispatched string,
and remove the localSlice field, hence QE should get the localSlice
from the map now. By this way, we avoid duplicating and modifying
the dispatch text string slice by slice, and each QE of a sliced
dispatch would get same contents now.

The extra space cost is sizeof(int) * SliceNumber bytes, and the extra
computing cost is iterating the SliceNumber-size array. Compared with
memcpy of text string for each slice in previous implementation, this
way is much cheaper, because SliceNumber is much smaller than the size
of dispatch text string. Also, since SliceNumber is so small, we just
use an array for the map instead of a hash table.

Also, clean up some dead code in dispatcher, including:
(1) Remove primary_gang_id field of Slice struct and DispatchCommandDtxProtocolParms
struct, since dispatch agent is deprecated now;
(2) Remove redundant logic in cdbdisp_dispatchX;
(3) Clean up buildGpDtxProtocolCommand;

### CODE REVIEW

File: src/backend/tcop/postgres.c [hunk 1]
```diff
@@ -4850,8 +4844,32 @@ PostgresMain(int argc, char *argv[],
...
+
+						for (localSlice = 0; localSlice < numSlices; ++localSlice)
+						{
+							if (sliceIndexGangIdMap[localSlice] == qe_gang_id)
+								break;
+						}
+
+						pfree(sliceIndexGangIdMap);
+					}
+
```
Comment (hlinnaka) [inline src/backend/tcop/postgres.c]: If the slice isn't found in the loop above, localSlice is left uninitialized. Shouldn't happen, of course, but let's check it to be sure.

**Comment by hlinnaka** [comment]
I like the cleanups and whitespace fixes, thanks for that. Might be good to put them in a separate commit, to make it easier to review the meaningful changes.

Hmm. I don't think there was anything wrong with old way of sending the slice id to execute, as part of the message. Admittedly, the way the message was constructed and dispatched in the QD was ugly, and I agree we should do something about that, but I think the protocol itself was fine. It seems simpler that the QD tells the QE "This is the plan, and this is the slice of that plan that you're going to execute", rather than "This is the plan, and this is how I mapped all the slices of the plan to gangs, and I told you earlier which gang you belong to".

So, how could we refactor the dispatching code, to get rid of the ugliness of duplicating and modifying the already-constructed protocol message? My first thought is to merge PQbuildGpQueryString() and PQsendGpQuery_shared() into one function, PQsendGpQuery(), that builds the protocol message, and sends it. Each dispatcher thread would just pass different slice ID parameter. It would be more like PQsendQuery() and other upstream libpq functions - PQsendGpQuery_shared() is a modularity violation as it is, because the caller needs to know about the protocol to construct the string correctly - that would become cleaner if the functions were merged.

That throws away the optimization of only building the message string once, and reusing the same buffer to all QEs. But I guess we had forgotten about that, anyway, as we modified the buffer to inject the slice ID anyway. This PR would (re-)enable that optimization, but I have to wonder how significant that really is. Code readability is probably more important here. And if we really wanted to, we could teach the new PQsendGpQuery() function to not memcpy() the pieces of the message to the output buffer, and avoid memory copying that way.

So, I'd suggest splitting of the primary_gang_id removal and the other cleanups to a separate patch, and get that committed first. And then explore making the dispatching code simpler, without changing the protocol.

**Comment by eurekaka** [comment]
Hi Heikki,

Thanks for the comments and suggestion.

> This PR would (re-)enable that optimization, but I have to wonder how significant that really is. 

The number of slices would grow with the growth of plan tree size, so SliceNumber * PlanTreeSize would be pretty large when query is complicated.

> And if we really wanted to, we could teach the new PQsendGpQuery() function to not memcpy() the pieces of the message to the output buffer, and avoid memory copying that way.

Sorry I don't understand how to avoid memcpy here, since we have to build a text string for dispatching just like how we do in PQbuildGpQueryString. If we build text string in thread, the cost should be larger than the current master branch implementation, since each thread would have a copy of the text now, while in master branch, each slice has a copy, and for this patch, there is only one copy.

Actually, I thought about hacking libpq to avoid memcpy then, but gave up for the following reasons:

- the text string would be shared by multiple connections, based on the current PQsendGpQuery_shared, it is hard to append a specific "localSlice" to the common part if the socket is set non-blocking;
- if we let libpq to construct text string and send it, we have to pass all the info in DispatchCommandQueryParms down to libpq level, this would make dispatcher code tightly coupled with libpq, and decrease the code readability. We want to limit the change to be within dispatcher, which keeps upstream merge simpler maybe, since we may remove gp_libpq_fe/ later and only keep one libpq version if we abandon threaded dispatching.
- dispatching same content for all slices may benefit potential features such as dispatch agent or broadcast dispatch.

**Comment by hlinnaka** [comment]
On 08/06/16 10:19, Kenan Yao wrote:
>> This PR would (re-)enable that optimization, but I have to wonder
>> how significant that really is.
>
> The number of slices would grow with the growth of plan tree size, so
> SliceNumber * PlanTreeSize would be pretty large when query is
> complicated.

Sure, but you have to compare it with the network traffic and the cost 
of actually executing the complex query. I don't think the memcpy() is 
usually very significant, in comparison. I'm sure you can construct a 
worst-case scenario query where it is, but I don't think we've had 
complaints from users on dispatch being slow because it consumes a lot 
of CPU or memory. Might be worth doing some performance testing to be 
sure, though.

>> And if we really wanted to, we could teach the new PQsendGpQuery()
>> function to not memcpy() the pieces of the message to the output
>> buffer, and avoid memory copying that way.
>
> Sorry I don't understand how to avoid memcpy here, since we have to
> build a text string for dispatching just like how we do in
> PQbuildGpQueryString. If we build text string in thread, the cost
> should be larger than the current master branch implementation, since
> each thread would have a copy of the text now, while in master
> branch, each slice has a copy, and for this patch, there is only one
> copy.

I'm thinking that instead of building the whole message into one linear 
buffer, you could for example, build a linked list of "fragments" that 
should be sent one after another, to make up the whole message. Some of 
the fragments could point to a shared buffer, for those parts of the 
message that is the same for every QE, while other fragments would be 
different. PQsendSome() would send all the fragments, one after another.

That same mechanism could be used to avoid the memcpy()s in e.g. 
PQsendQuery() as well (that doens't matter much for GPDB, though, 
because we don't use it for dispatching from QD to QEs). It would be an 
isolated change in libpq code.

> Actually, I thought about hacking libpq to avoid memcpy then, but
> gave up for the following reasons:
>
> - the text string would be shared by multiple connections, based on
> the current PQsendGpQuery_shared, it is hard to append a specific
> "localSlice" to the common part if the socket is set non-blocking; -
> if we let libpq to construct text string and send it, we have to pass
> all the info in DispatchCommandQueryParms down to libpq level, this
> would make dispatcher code tightly coupled with libpq, and decrease
> the code readability.

It seems OK to me. DispatchCommandQueryParms could well be part of the 
"public" API of gp_libpq. It's similar to other libpq functions, like 
PQexecParams(), for example. The caller passes the query string, and the 
parameters, and PQexecParams() builds the message to send over the wire. 
Same here: the caller would pass the SQL command, the plan tree, the 
parameters, and so forth, to PQsendGpQuery(), which builds the message 
to send over the wire.

There's some room for argument on whether we PQsendGpQuery() should take 
DispatchCommandQueryParms struct as argument, or if each of those thing, 
the SQL command, the plan tree, etc., should be separate arguments. 
Having them as separate arguments would be closer to how PQexecParams() 
and other libpq functions work, but it might be a bit cumbersome because 
the function would have so many arguments.

Either way, it would be better from readability and modularity point of 
view if the dispatcher code didn't need to know the details of the FE/BE 
protocol.

> We want to limit the change to be within
> dispatcher, which keeps upstream merge simpler maybe, since we may
> remove gp_libpq_fe/ later and only keep one libpq version if we
> abandon threaded dispatching.

Agreed. I don't see big differences on that aspect, though, regardless 
of which approach we choose. Dealing with the "fragments" rather than a 
single buffer in PQsendSome() would probably be more code, but it would 
be in the same place as the current modifications in libpq. So it 
wouldn't really increase the "diff footprint".

  - dispatching same content for all
> slices may benefit potential features such as dispatch agent or
> broadcast dispatch.

Well, that's a whole different story :-). Broadcast would be so 
different that I don't think any of this matters for that.

- Heikki

**Comment by eurekaka** [comment]
> I'm thinking that instead of building the whole message into one linear
buffer, you could for example, build a linked list of "fragments" that
should be sent one after another, to make up the whole message. Some of
the fragments could point to a shared buffer, for those parts of the
message that is the same for every QE, while other fragments would be
different. PQsendSome() would send all the fragments, one after another.

Thanks, this is a great idea, but we have to hack pg_conn struct and pqSendSome(), which is pretty inside libpq.

> Either way, it would be better from readability and modularity point of view if the dispatcher code didn't need to know the details of the FE/BE protocol.

Hmm, actually PQbuildGpQueryString should be a public API of libpq(I would move it to libpq files), so now dispatcher is calling API of libpq to build text string, and it does not know FE/BE protocol indeed, so it seems OK to me from the readability and modularity point of view.

> Dealing with the "fragments" rather than a single buffer in PQsendSome() would probably be more code, but it would be in the same place as the current modifications in libpq. So it wouldn't really increase the "diff footprint".

But now we have an alternative option here to avoid increase diffs.

Frankly speaking, taking performance, readability, modularity and continuous merge all into consideration, the approach in this patch and the approach by hacking libpq to append 'localSlice' are half and half to me. Since we have this patch already, I would prefer to use this one.

File: src/backend/tcop/postgres.c [hunk 2]
```diff
@@ -4850,8 +4844,32 @@ PostgresMain(int argc, char *argv[],
 					if (seqServerHostlen > 0)
 						seqServerHost = pq_getmsgbytes(&input_message, seqServerHostlen);
 
+					numSlices = pq_getmsgint(&input_message, 4);
```
Comment (xiong-gang) [inline src/backend/tcop/postgres.c]: I think it's better to do it in one loop and error out if 'localslice' is not found

**Comment by xiong-gang** [comment]
LGTM in general.

**Comment by eurekaka** [comment]
pushed with comments resolved, thanks for reviewing.

**Pull request closed without merging**